homeassistant/components/snmp/device_tracker.py

```diff
@@ -51,78 +71,119 @@ class SnmpScanner(DeviceScanner):
     """Queries any SNMP capable Access Point for connected devices."""
 
     def __init__(self, config):
-        """Initialize the scanner."""
-
-        self.snmp = cmdgen.CommandGenerator()
-
-        self.host = cmdgen.UdpTransportTarget((config[CONF_HOST], 161))
-        if CONF_AUTH_KEY not in config or CONF_PRIV_KEY not in config:
-            self.auth = cmdgen.CommunityData(config[CONF_COMMUNITY])
+        """Initialize the scanner and test the target device."""
+        host = config.get(CONF_HOST)
+        port = DEFAULT_PORT
+        community = config.get(CONF_COMMUNITY)
+        baseoid = config.get(CONF_BASEOID)
+        version = DEFAULT_VERSION
+        username = community
+        authkey = config.get(CONF_AUTH_KEY)
+        authproto = DEFAULT_AUTH_PROTOCOL
+        privkey = config.get(CONF_PRIV_KEY)
+        privproto = DEFAULT_PRIV_PROTOCOL
+
+        try:
+            # Try IPv4 first.
+            target = UdpTransportTarget((host, port), timeout=DEFAULT_TIMEOUT)
+        except PySnmpError:
+            # Then try IPv6.
+            try:
+                target = Udp6TransportTarget((host, port), timeout=DEFAULT_TIMEOUT)
+            except PySnmpError as err:
+                _LOGGER.error("Invalid SNMP host: %s", err)
+                return
+
+        if authkey is not None or privkey is not None:
+            if not authkey:
+                authproto = "none"
+            if not privkey:
+                privproto = "none"
+
+            request_args = [
+                SnmpEngine(),
+                UsmUserData(
+                    username,
+                    authKey=authkey or None,
+                    privKey=privkey or None,
+                    authProtocol=authproto,
+                    privProtocol=privproto,
+                ),
+                target,
+                ContextData(),
+            ]
         else:
-            self.auth = cmdgen.UsmUserData(
-                config[CONF_COMMUNITY],
-                config[CONF_AUTH_KEY],
-                config[CONF_PRIV_KEY],
-                authProtocol=cfg.usmHMACSHAAuthProtocol,
-                privProtocol=cfg.usmAesCfb128Protocol,
-            )
-        self.baseoid = cmdgen.MibVariable(config[CONF_BASEOID])
+            request_args = [
+                SnmpEngine(),
+                CommunityData(community, mpModel=SNMP_VERSIONS[version]),
+                target,
+                ContextData(),
+            ]
+
+        self.request_args = request_args
+        self.baseoid = baseoid
         self.last_results = []
+        self.success_init = False
 
-        # Test the router is accessible
-        data = self.get_snmp_data()
+    async def test_device_readable(self):
```

Comment:
```suggestion
    async def async_test_device_readable(self):
```

As we have both sync and async code in this codebase, please prefix any function safe run in the event loop with `async_`